File: src/mvmctl/api/inputs/_config_input.py
"""Config input resolution — Input → Request → ResolvedConfigInput."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from mvmctl.core._shared import Database
from mvmctl.core.config._repository import SettingsRepository
from mvmctl.core.config._service import OVERRIDABLE_SETTINGS, SettingsService
from mvmctl.exceptions import ConfigError


@dataclass
class ConfigInput:
    """Raw config parameters from CLI."""

    action: str  # 'get', 'set', 'list', 'reset'
    category: str | None = None  # e.g. 'defaults.vm'
    key: str | None = None  # e.g. 'vcpu_count'
    value: Any | None = None  # for 'set'
    all_overrides: bool = False  # for 'reset --all'


@dataclass(frozen=True)
class ResolvedConfigInput:
    """Fully resolved config operation parameters."""

    action: str
    category: str | None
    key: str | None
    value: Any | None
    all_overrides: bool
    service: SettingsService


class ConfigRequest:
    """Resolve ConfigInput against the database."""

    def __init__(self, inputs: ConfigInput, db: Database | None = None) -> None:
        self._inputs = inputs
        self._db = db if db is not None else Database()
        self._service = SettingsService(SettingsRepository(self._db))

    def resolve(self) -> ResolvedConfigInput:
        """Resolve and validate config input."""
        category = self._inputs.category
        key = self._inputs.key

        if self._inputs.action == "get":
            if not category:
                raise ConfigError("Category is required for get operation")
            # key is optional for category-level get
            if key is not None:
                cat_settings = OVERRIDABLE_SETTINGS.get(category)
                if cat_settings is None or key not in cat_settings:
                    raise ConfigError(
                        f"'{category}.{key}' is not a valid setting key. "
                        f"Use 'mvm config ls' to see valid keys."
                    )

        elif self._inputs.action == "set":
            if not category or not key:
                raise ConfigError(
                    "Category and key are required for set operation"
                )
            if self._inputs.value is None:
                raise ConfigError("Value is required for set operation")

            # Validate key is overridable (caller validates, receiver trusts)
            cat_settings = OVERRIDABLE_SETTINGS.get(category)
            if cat_settings is None or key not in cat_settings:
                raise ConfigError(
                    f"'{category}.{key}' is not an overridable setting. "
                    f"Use 'mvm config ls' to see valid keys."
                )

        elif self._inputs.action == "reset":
            if self._inputs.all_overrides:
                # category and key are both optional for --all
                pass
            elif not category:
                raise ConfigError(
                    "Category is required for reset operation (or use --all)"
                )
            # key is optional for category-level reset
            if key is not None:
                cat_settings = OVERRIDABLE_SETTINGS.get(category or "")
                if cat_settings is None or key not in cat_settings:
                    raise ConfigError(
                        f"'{category}.{key}' is not a valid setting key"
                    )

        return ResolvedConfigInput(
            action=self._inputs.action,
            category=category,
            key=key,
            value=self._inputs.value,
            all_overrides=self._inputs.all_overrides,
            service=self._service,
        )

File: pkg/api/inputs/config_input.go
package inputs

import (
	"mvmctl/internal/core/config"
	"mvmctl/pkg/errs"
)

// ConfigInput specifies config input.
type ConfigInput struct {
	Action       string `json:"action"`
	Category     string `json:"category,omitempty"`
	Key          string `json:"key,omitempty"`
	Value        any    `json:"value,omitempty"`
	AllOverrides bool   `json:"all_overrides"`
}

// Validate checks that the config input is valid for the given action.
func (i *ConfigInput) Validate() error {
	switch i.Action {
	case "get":
		if i.Category == "" {
			return errs.New(errs.CodeConfigError, "Category is required for get operation")
		}
		if i.Key != "" && !config.IsKeyInCategory(i.Category, i.Key) {
			return errs.New(
				errs.CodeConfigError,
				"'"+i.Category+"."+i.Key+"' is not a valid setting key. Use 'mvm config ls' to see valid keys.",
			)
		}
	case "set":
		if i.Category == "" || i.Key == "" {
			return errs.New(errs.CodeConfigError, "Category and key are required for set operation")
		}
		if i.Value == nil {
			return errs.New(errs.CodeConfigError, "Value is required for set operation")
		}
		if !config.IsKeyInCategory(i.Category, i.Key) {
			return errs.New(
				errs.CodeConfigError,
				"'"+i.Category+"."+i.Key+"' is not an overridable setting. Use 'mvm config ls' to see valid keys.",
			)
		}
	case "reset":
		if i.AllOverrides {
			// category and key are both optional for --all
		} else if i.Category == "" {
			return errs.New(errs.CodeConfigError, "Category is required for reset operation (or use --all)")
		}
		if i.Key != "" && !config.IsKeyInCategory(i.Category, i.Key) {
			return errs.New(errs.CodeConfigError, "'"+i.Category+"."+i.Key+"' is not a valid setting key")
		}
	case "list":
		// No additional validation needed for listing
	default:
		return errs.New(errs.CodeConfigError, "Invalid action: '"+i.Action+"'. Use get, set, reset, or list.")
	}
	return nil
}

// Resolve validates the config input and returns it. The resolved shape is
// identical to the input, so this method simply validates and returns self.
func (i *ConfigInput) Resolve() (*ConfigInput, error) {
	if err := i.Validate(); err != nil {
		return nil, err
	}
	return i, nil
}
